aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -443,22 +477,42 @@ module AptosFramework::Stake {
         // Officially deactivate all pending_inactive validators. They will now no longer receive rewards.
         validator_set.pending_inactive = Vector::empty();
 
-        // Update validator info (so network address/public key change takes effect)
+        // Reset performance scores and update active validator set (so network address/public key change takes effect).
         let i = 0;
         let len = Vector::length(&validator_set.active_validators);
         let active_validators = Vector::empty();
+        validator_perf.num_blocks = 1;
+        validator_perf.missed_votes = Vector::empty();
         while (i < len) {
             let old_validator_info = Vector::borrow_mut(&mut validator_set.active_validators, i);
-            let new_validator_info = generate_validator_info(old_validator_info.addr);
+            let pool_address = old_validator_info.addr;
+            let validator_config = borrow_global_mut<ValidatorConfig>(pool_address);
+            let new_validator_info = generate_validator_info(pool_address, *validator_config);
             if (new_validator_info.voting_power >= validator_set.minimum_stake &&
                 new_validator_info.voting_power <= validator_set.maximum_stake
             ) {
                 Vector::push_back(&mut active_validators, new_validator_info);
+                Vector::push_back(&mut validator_perf.missed_votes, 0);
             };
             i = i + 1;
         };
 
+        // Sort the validator set, so we can ensure stable order and validator indices.
         sort_validators(&mut active_validators);
+
+        // Store the validator index in the validator's ValidatorConfig resource, so we can look it up later.
+        // We can't rely on the validator set in CoreResource's ValidatorSet because some validators might choose
+        // to leave and get moved to pending_inactive later on.
+        let i = 0;
+        let len = Vector::length(&active_validators);
+        while (i < len) {
+            let validator_info = Vector::borrow(&active_validators, i);
+            let pool_address = validator_info.addr;
+            let validator_config = borrow_global_mut<ValidatorConfig>(pool_address);
+            validator_config.validator_index = i;
+            i = i + 1;
+        };
```

Comment:
so ugly

Reply:
@zekun000 and I discussed this - we actually no longer need the validator set to be sorted as the validator index is now stable and easily accessible in ValidatorSet/ValidatorConfig. So the block_executor code can actually compute the vote bitmap based on these indices directly and thus ensure the vote bitmap would have the same order as the (unsorted) validator set stored on chain.

@zekun000 mentioned he'll look into implementing this later as it's potentially blocking some other work he's been working on.